### PR TITLE
Revise host mode.

### PR DESCRIPTION
**Pull request by pgavlin**
The current implementation of host mode uses a `pulumi host` command and
an ad-hoc communication protocol between the engine and client to
connect a language host after the host has begun listening. The most
significant disadvantages of this approach are the communication
protocol (which currently requires the use of stdout), the host-specific
command, and the difficulty of accommodating the typical program-bound
lifetime for an update.

These changes reimplement host mode by adding engine support for
connecting to an existing language runtime service rather than launching
a plugin. This capability is provided via an engine-specific language
runtime, `client`, which accepts the address of the existing languge
runtime service as a runtime option. The CLI exposes this runtime via
the `--client` flag to the `up` and `preview` commands, which similarly
accepts the address of an existing language runtime service as an
argument. These changes also adjust the automation API to consume the
new host mode implementation.

### CODE REVIEW

**Review by justinvp — APPROVED** [review]
LGTM. Would be good for @EvanBoyle to review as well.

**Comment by EvanBoyle** [comment]
Just linking these two issues:

Closes https://github.com/pulumi/pulumi/issues/5225
closes https://github.com/pulumi/pulumi/issues/5249

File: sdk/go/common/resource/plugin/host.go [hunk 1]
```diff
@@ -73,8 +73,6 @@ type Host interface {
 	// EnsurePlugins ensures all plugins in the given array are loaded and ready to use.  If any plugins are missing,
 	// and/or there are errors loading one or more plugins, a non-nil error is returned.
 	EnsurePlugins(plugins []workspace.PluginInfo, kinds Flags) error
-	// GetRequiredPlugins lists a full set of plugins that will be required by the given program.
-	GetRequiredPlugins(info ProgInfo, kinds Flags) ([]workspace.PluginInfo, error)
```
Comment (EvanBoyle) [inline sdk/go/common/resource/plugin/host.go:77]: Do you have any ideas on how we could integrate automatic plugin acquisition with the new world? Today, it's annoying but I think most users will be able to bear. As multi-lang and component consumption become more common, hand managing dependencies in this execution mode becomes problematic.

**Review by EvanBoyle — APPROVED** [review]
This looks great. Substantially cleaner and easier to reason about than the original approach. Thanks for taking the time to dig into this! Just a few notes/questions.

File: pkg/cmd/pulumi/util.go [hunk 2]
```diff
@@ -399,6 +399,24 @@ func parseAndSaveConfigArray(s backend.Stack, configArray []string, path bool) e
...
+// containing directory, which will be used as the root of the project's Pulumi program. If a
+// client address is present, the returned project will always have the runtime set to "client"
+// with the address option set to the client address.
+func readProjectForUpdate(clientAddress string) (*workspace.Project, string, error) {
+	proj, root, err := readProject()
+	if err != nil {
+		return nil, "", err
+	}
+	if clientAddress != "" {
+		proj.Runtime = workspace.NewProjectRuntimeInfo("client", map[string]interface{}{
```
[the remark this answers is not in the document]
Comment (pgavlin) [reply pkg/cmd/pulumi/util.go:413]: I'm going to leave this as-is for now and address it in a follow-up change. I'll file an issue to track.

File: sdk/go/common/resource/plugin/langruntime_plugin.go [hunk 3]
```diff
@@ -156,14 +143,6 @@ func (h *langhost) GetRequiredPlugins(info ProgInfo) ([]workspace.PluginInfo, er
 func (h *langhost) Run(info RunInfo) (string, bool, error) {
 	logging.V(7).Infof("langhost[%v].Run(pwd=%v,program=%v,#args=%v,proj=%s,stack=%v,#config=%v,dryrun=%v) executing",
 		h.runtime, info.Pwd, info.Program, len(info.Args), info.Project, info.Stack, len(info.Config), info.DryRun)
-	// In host mode, we simply print the engine and monitor address to stderr and wait for a signal.
-	if h.IsHostOnlyMode() {
-		fmt.Fprintf(os.Stderr, "resmon: %s\n", info.MonitorAddress)
-		sigs := make(chan os.Signal, 1)
-		signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
```
[the remark this answers is not in the document]
Comment (pgavlin) [reply sdk/go/common/resource/plugin/langruntime_plugin.go:163]: Awesome; good catch!

File: sdk/go/common/resource/plugin/host.go (hunk 1 above)
Comment (pgavlin) [reply sdk/go/common/resource/plugin/host.go:77]: In principle all we need to do is implement `GetRequiredPlugins` in each SDK's language service. Ideally we'd factor out the code that's currently used to do this, but it's not always in the SDK's target language (e.g. the implementation of the Node language provider is currently a Go binary that launches Node in response to `Run`).

File: pkg/engine/lifecycle_test.go [hunk 4]
```diff
@@ -5990,3 +5994,112 @@ func TestSingleComponentDefaultProviderLifecycle(t *testing.T) {
...
+
+func startUpdate(host plugin.Host) (*updateContext, error) {
+	ctx := &updateContext{
+		resmon:       make(chan *deploytest.ResourceMonitor),
+		programErr:   make(chan error),
+		snap:         make(chan *deploy.Snapshot),
+		updateResult: make(chan result.Result),
+	}
+
+	port, _, err := rpcutil.Serve(0, nil, []func(*grpc.Server) error{
```
[the remark this answers is not in the document]
Comment (pgavlin) [reply pkg/engine/lifecycle_test.go:6015]: That's a good idea, yes.

File: pkg/cmd/pulumi/preview.go [hunk 5]
```diff
@@ -195,6 +196,9 @@ func newPreviewCmd() *cobra.Command {
 		&configPath, "config-path", false,
 		"Config keys contain a path to a property in a map or list to set")
 
+	cmd.PersistentFlags().StringVar(
+		&client, "client", "", "The address of an existing language runtime host to connect to")
```
[the remark this answers is not in the document]
Comment (pgavlin) [reply pkg/cmd/pulumi/preview.go:200]: Sure.

File: pkg/cmd/pulumi/util.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (EvanBoyle) [reply pkg/cmd/pulumi/util.go:413]: Opened https://github.com/pulumi/home/issues/1089

**Pull request merged**